File: src/0442-Find-All-Duplicates-in-an-Array/find_all_duplicates_in_an_array.cpp
#include <iostream>
#include <vector>
#include <cmath>

class Solution {
public:
    std::vector<int> findDuplicates(std::vector<int>& nums) {
        std::vector<int> ans = {};

        for(int i = 0; i < nums.size(); i++) {
            int pos = std::abs(nums[i]) - 1;

            if (nums[pos] < 0) {
                ans.push_back(pos + 1);
            }

            nums[pos] = -nums[pos];
        }

        return ans;
    }
};

int main() {
    std::vector<int> nums = {1, 3, 4, 2, 2};
    Solution solution;
    
    std::vector<int> output = solution.findDuplicates(nums);

    for (auto it : output) {
        std::cout << it << " ";
    }

    return 0;
}

File: src/0442-Find-All-Duplicates-in-an-Array/find_all_duplicates_in_an_array.py
from typing import List

class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        ans: List[int] = []

        for i in range(len(nums)):
            pos: int = abs(nums[i]) - 1

            if nums[pos] < 0:
                ans.append(pos + 1)
                
            nums[pos] = -nums[pos]

        return ans

if __name__ == "__main__":
    nums: List[int] = [4,3,2,7,8,2,3,1]
    solution = Solution()
    
    print(solution.findDuplicates(nums))